src/lib/tuning/index/index_evaluator.cpp

```diff
@@ -0,0 +1,76 @@
+#include "index_evaluator.hpp"
+
+#include "optimizer/column_statistics.hpp"
+#include "optimizer/table_statistics.hpp"
+#include "resolve_type.hpp"
+#include "storage/index/base_index.hpp"
+#include "storage/storage_manager.hpp"
+#include "storage/table.hpp"
+#include "types.hpp"
+#include "utils/logging.hpp"
+
+namespace opossum {
+
+IndexEvaluator::IndexEvaluator() {}
+
+void IndexEvaluator::_setup() { _saved_work.clear(); }
+
+void IndexEvaluator::_process_access_record(const BaseIndexEvaluator::AccessRecord& record) {
+  const auto table_statistics = StorageManager::get().get_table(record.column_ref.table_name)->table_statistics();
+  // ToDo(anyone) adapt for multi column indices...
+  const auto predicate_statistics =
+      table_statistics->predicate_statistics(record.column_ref.column_ids[0], record.condition, record.compare_value);
+  const auto total_rows = table_statistics->row_count();
+  const auto match_rows = predicate_statistics->row_count();
+  const auto unscanned_rows = total_rows - match_rows;
+  const float saved_work = unscanned_rows * record.query_frequency;
+  LOG_INFO("saved work for query on " << record.column_ref.table_name << "." << record.column_ref.column_ids[0] << ": "
+                                      << saved_work << "\n");
+  if (_saved_work.count(record.column_ref) > 0) {
+    _saved_work[record.column_ref] += saved_work;
+  } else {
+    _saved_work[record.column_ref] = saved_work;
+  }
+}
+
+ColumnIndexType IndexEvaluator::_propose_index_type(const IndexChoice& index_evaluation) const {
+  return ColumnIndexType::GroupKey;
+}
+
+uintptr_t IndexEvaluator::_predict_memory_cost(const IndexChoice& index_evaluation) const {
+  const auto table = StorageManager::get().get_table(index_evaluation.column_ref.table_name);
+  // ToDo(anyone) adapt for multi column indices...
+  const auto column_statistics =
+      table->table_statistics()->column_statistics().at(index_evaluation.column_ref.column_ids[0]);
+  const auto distinct_value_count = column_statistics->distinct_count();
+
+  // Sum up column data type widths
+  size_t value_bytes = 0;
+  for (const auto column_id : index_evaluation.column_ref.column_ids) {
```

Comment:
Five lines above, you seem to assume that there is only one column id

Reply:
Six lines above we commented a ToDo to add multi-column support. This is currently not useful as there are no multi-column scans that would use a multi-column index. We tried to make our code multi-column aware where this meant no additional effort, but did not bother adapting the system where it clearly did not support multi-columns. This is the case with `TableStatistics` where instead of adapting it we added the note to adapt this line in a future where multi-column indexes are relevant.

If you think consistency is more important than preparations for the future, we will remove this loop and assume/assert `column_ids.size()==0`.

Reply:
Missed that comment. We should assert that there is only one column, but that doesn't mean that we can't keep the loop for the future.

Reply:
Added a DebugAssert